dulwich/sparse_patterns.py

```diff
@@ -0,0 +1,462 @@
+            included.add(path_str)
+
+    return included
+
+
+def apply_included_paths(repo, included_paths, force=False):
+    """Apply the sparse-checkout inclusion set to the index and working tree.
+
+    This function updates skip-worktree bits in the index based on whether each
+    path is included or not. It then adds or removes files in the working tree
+    accordingly. If ``force=False``, files that have local modifications
+    will cause an error instead of being removed.
+
+    Args:
+      repo: A path to the repository or a Repo object.
+      included_paths: A set of paths (strings) that should remain included.
+      force: Whether to forcibly remove locally modified files (default False).
+
+    Returns:
+      None
+    """
+    index = repo.open_index()
+    normalizer = repo.get_blob_normalizer()
+
+    def local_modifications_exist(full_path, index_entry):
+        if not os.path.exists(full_path):
+            return False
+        try:
+            with open(full_path, "rb") as f:
+                disk_data = f.read()
+        except OSError:
+            return True
+        try:
+            blob = repo.object_store[index_entry.sha]
+        except KeyError:
+            return True
+        norm_data = normalizer.checkin_normalize(disk_data, full_path)
+        return norm_data != blob.data
+
+    # 1) Update skip-worktree bits
+    for path_bytes, entry in list(index.items()):
+        path_str = path_bytes.decode("utf-8")
+        if path_str in included_paths:
+            entry.set_skip_worktree(False)
+        else:
+            entry.set_skip_worktree(True)
+        index[path_bytes] = entry
+    index.write()
+
+    # 2) Reflect changes in the working tree
+    for path_bytes, entry in list(index.items()):
+        full_path = os.path.join(repo.path, path_bytes.decode("utf-8"))
+
+        if entry.skip_worktree:
+            # Excluded => remove if safe
+            if os.path.exists(full_path):
+                if not force and local_modifications_exist(full_path, entry):
+                    raise CheckoutError(
+                        f"Local modifications in {full_path} would be overwritten "
+                        "by sparse checkout. Use force=True to override."
+                    )
+                try:
+                    os.remove(full_path)
+                except IsADirectoryError:
+                    pass
+                except FileNotFoundError:
+                    pass
+        else:
+            # Included => materialize if missing
+            if not os.path.exists(full_path):
+                try:
+                    blob = repo.object_store[entry.sha]
+                except KeyError:
+                    raise Error(f"Blob {entry.sha} not found for {path_bytes}.")
+                ensure_dir_exists(os.path.dirname(full_path))
+                with open(full_path, "wb") as f:
+                    f.write(blob.data)
+
+
+class ParsedPattern:
```

Comment:
Please reuse the existing gitignore pattern code.

Reply:
this bit was draft code that wasn't being called sorry, should've been cut before pushing